Use Launcher App for Launch at Login

nidegen opened the pull request:
Hi Steve,

As discussed a few weeks ago, there is the way to launch the app on login with a launch helper app.

Basically, you can only register an App to launch at login from outside the app itself, hence we need a helper app that is registered and doesn't do anything but launch the main app when it is started.

I tried to remove IYLoginItems completely but as I have no experience with CocoaPods, please check if I didn't mess up anything.

Let me know if you can test it succesfully:)

Nicolas

SteveBarnegren commented:
Amazing work, thanks!

I'll be away for most of the weekend, but I'll test either towards the end of the weekend or early next week, then hopefully I can get it uploaded to the App Store!

SteveBarnegren commented:
Hey Nicolas!

It all looks good, but I can't quite get it to work! Here's what I'm doing:

- In the current version of TopDrawer, uncheck _'open at login'_ in the settings, so that it's removed from the login items. (so we don't get a false positive)
- Build the new version, and overwrite the old one in the applications directory
- Open the new TopDrawer and check _'open at login'_ in the settings
- Restart machine
- TopDrawer doesn't launch

Took a quick look at the code, and I think there's a couple of things that might be missing.

1. There should be some point at which the helper application gets registered to be run on launch somehow. I can't see where that is at the moment.
2. When toggling the _'open at login'_ checkbox in the settings panel, that results in `AppDelegate.launchAtLoginUserDefaultKey` being set in `UserDefaults`, but I think the helper application should probably read that value when it launches and then decide not to launch the main application if it's `false`.

Let me know if you can get it working and I'll give it another test. Or if you think that I'm not testing it correctly then let me know!

nidegen commented:
Hi Steve,

Yeah you are right of course. I was a bit sloppy as I thought I'd first push a first draft for review. In the rush I thought it would be all implemented and straightforward - sorry for that.

I added the missing line and fixed a typo I noticed when testing it after all. For my machine I can now assert it runs as expected. Let me know if that fixed it on your side too!

nidegen commented:
sorry the first PR wasn't where it should have been, but this should be fine now:)

No hurry, but just wanted to make sure it doesn't get lost;)

SteveBarnegren commented:
Hi Nicolas,

Sorry for the delay in getting back to you. I did test it, I've just didn't get round to commenting back on the PR, as I've been pretty busy for the last couple of weeks!

The changes to the code that you've made look good to me, although I was still unable to get it to open at login, again following the steps above.

Can you post the steps that you're following to test it, and I'll see if I can do the same?

SteveBarnegren commented on TopDrawer/Settings/Settings.swift:
```diff
@@ -115,6 +115,8 @@ class Settings {
         self.path = Setting(keyValueStore: keyValueStore, key: "path", defaultValue: "")
         self.shortenPaths = Setting(keyValueStore: keyValueStore, key: "shortenPaths", defaultValue: true)
+        self.launchAtLogin = Setting(keyValueStore: keyValueStore, key: AppDelegate.launchAtLoginUserDefaultKey, defaultValue: true)
+        SMLoginItemSetEnabled(AppDelegate.launcherAppId as CFString, true)
```
Looks like there's a bug here. When `Settings` is initialised the `SMLoginItemSetEnabled(true)` is set, which would set the application to open at login every time it's run, even if the user has disabled it?

I think the value passed to `SMLoginItemSetEnabled()` should be the current open at login state. So `SMLoginItemSetEnabled(self.launchAtLogin.value)`

nidegen replied:
yep, true. That always made it launch on login. Actually we dont need to sync it at all as SMLoginItem are persistently stored anyways.

For me it now works fine, enabling or disabling the option.

I do compile the app, launch the .app bundle that lays in the Product folder. From there, when I log out and log in again, it works as expected. Also when disabled now:)

nidegen replied:
I also tested it when moving to the Application folder. Works as well..

SteveBarnegren commented:
Hi Nicolas,

Thanks for making those changes!

There may have some issues with testing on my setup, because I was still on High Sierra. I've updated to Mojave now, so that might resolve things. I don't think Top-Drawer needs to support anything but the latest macOS anyway, so If it works in Mojave then I'm happy with that.

I'm going to be travelling for the next couple of weeks, so I'll probably have to pick this up after that. When I'm back I'll give it another test though, and see if I can get it submitted to the store soon after.

nidegen commented:
Hi Steve,

Hmm, no I guess the same approach was working on High Sierra as well. I will try to double check everything then.

Enjoy your travels!

Nicolas

SteveBarnegren commented:
@nidegen Sorry for the delay in getting around to testing this again. Happy to confirm it all worked fine for me this time!

I'm not sure if it made any difference that I've updated to Mojave since the last time that I tested or not. Either way, I'm going to raise the minimum version to Mojave now, because I don't have a good way to test on older versions of macOS, and it's probably not worth the additional time investment anyway.

Thanks so much for all of the time that you've put in to this! I'll merge this in now, and I'll try to get a build submitted to the App Store in the next week or so!
